### ISSUE TITLE
Investigate and Rationalise use of SSM Resource Data Syncs on Mod Platform

### ISSUE THREAD

**Issue by richgreen-moj** [ministryofjustice/modernisation-platform#8624]
### User Story

As a MP Engineer
I need to investigate the origin of the existing resource data syncs that are applied across the MoJ org/MP accounts
So that I can make the best use of them to gather inventory information for licensing or software information for Mod Platform accounts

### Value / Purpose

As part of issue https://github.com/ministryofjustice/modernisation-platform/issues/8360 we looked into deploying a ssm inventory resource data sync in the member bootstrap code with the aim of aggregating all the software deployed across the platform in one place (an s3 bucket in the `organisation-security` account) so that it could be queried using Athena.

During the work to deploy it we came across an issue where many of the MP accounts (at least 130) had already hit a hard limit of 5 resource data syncs so we were unable to deploy it successfully.

The purpose of this ticket is to see whether we can confirm whether any of these resource data syncs are in use by anyone in MoJ and where possible to delete any data syncs and associated infrastructure that are not in use.

### Context / Background

Most of the affected member accounts have the following data syncs:

| Data Sync Name  | Origin |
| ------------- | ------------- |
| InspectorResourceDataSync-do-not-delete | In all accounts and created by the AWS Inspector service which is turned on in baselines|
| license-manager-cross-account-discovery-for-del-admin-v2  | Targets a bucket we cannot see the origin of |
| license-manager-cross-account-discovery-for-del-admin1710e486e9d | Targets a bucket we can see in the `organisation-security` account  |
| license-manager-cross-account-discovery-v2 | Targets a bucket we cannot see the origin of |
| license-manager-resource-data-sync-for-cross-account-discovery | Targets a bucket we can see in the `moj-master` account   |

Some work has gone into looking at where these may have originated from. There aren't any references to the creation of the "license-manager" resource data syncs in any repos in the MoJ organisation that we can see apart from a PoC that was carried out [here](https://github.com/ministryofjustice/aws-root-account/blob/e52ef4e7b148499b5a30f69e4ae18a87d27ac7d2/organisation-security/terraform/license-manager.tf) and [here](https://github.com/ministryofjustice/aws-root-account/blob/e52ef4e7b148499b5a30f69e4ae18a87d27ac7d2/organisation-security/terraform/ssm.tf#L1-L115) which is being revisited in [another ticket](https://github.com/ministryofjustice/modernisation-platform/issues/8023).

### Useful Contacts

@richgreen-moj @ewastempel @davidkelliott

### Definition of Done

- [ ] Investigate further the origin of these data syncs and, where possible, identify candidates for deletion
- [ ] Carry out deletion of any unused data syncs and associated infrastructure

**Comment by ewastempel** [comment]
In order to see the pattern and understand it more what created the ssm resource data syncs at the first place, I have exported information on the ssm resource data sync in a couple of accounts (json format). These are attached here and will be studied as the next step.

[resource-data-sync-electronic-monitoring-data-preproduction-eu-west-1.txt](https://github.com/user-attachments/files/18802744/resource-data-sync-electronic-monitoring-data-preproduction-eu-west-1.txt)

[resource-data-sync-electronic-monitoring-data-preproduction-eu-west-2.txt](https://github.com/user-attachments/files/18802746/resource-data-sync-electronic-monitoring-data-preproduction-eu-west-2.txt)

[resource-data-sync-modernisation-platform-eu-west-1.txt](https://github.com/user-attachments/files/18802752/resource-data-sync-modernisation-platform-eu-west-1.txt)

[resource-data-sync-modernisation-platform-eu-west-2.txt](https://github.com/user-attachments/files/18802757/resource-data-sync-modernisation-platform-eu-west-2.txt)

[resource-data-sync-moj-master-eu-west-1.txt](https://github.com/user-attachments/files/18802759/resource-data-sync-moj-master-eu-west-1.txt)

[resource-data-sync-moj-master-eu-west-2.txt](https://github.com/user-attachments/files/18802761/resource-data-sync-moj-master-eu-west-2.txt)

[resource-data-sync-oasys-dev-eu-west-1.txt](https://github.com/user-attachments/files/18802765/resource-data-sync-oasys-dev-eu-west-1.txt)

[resource-data-sync-oasys-dev-eu-west-2.txt](https://github.com/user-attachments/files/18802766/resource-data-sync-oasys-dev-eu-west-2.txt)

[resource-data-sync-org-security-eu-west-1.txt](https://github.com/user-attachments/files/18802772/resource-data-sync-org-security-eu-west-1.txt)

[resource-data-sync-org-security-eu-west-2.txt](https://github.com/user-attachments/files/18802782/resource-data-sync-org-security-eu-west-2.txt)

**Comment by ewastempel** [comment]
Initially it looked like three of the ssm resource data syncs were in use and two of them could be deleted (see the attached picture for visualisation and better understanding). This was believed as when creating a new account, the three data syncs are provisioned with the accounts. One of the data syncs in use is dedicated to Inspector and it appears in both London and Ireland regions. The other two in use seem to relate to the License Manager. The other two that do not appear to be in use are also related by the License Manager.
In use:
```
InspectorResourceDataSync-do-not-delete
license-manager-cross-account-discovery-for-del-admin-v2
license-manager-cross-account-discovery-v2
```
Not in use:
```
license-manager-resource-data-sync-for-cross-account-discovery
license-manager-cross-account-discovery-for-del-admin1710e486e9d
```

However, later on, when looking in `organisation-security` and `MoJ Master` accounts, these two (from not in use list) have links to s3 resources that exist in both accounts:
```
license-manager-resource-data-sync-for-cross-account-discovery
license-manager-cross-account-discovery-for-del-admin1710e486e9d
```

Which in turn, suggests that they are in use. This discovery is contradicting and counter intuitive.

A support case was raised to get more clarity from AWS on how to best identify which resources are still in use and which are not and also to ask for the best approach in order to clean up such resources across all of the accounts.

The support case reference raised in the org security account: **Case ID 174076546100711**

**Comment by ewastempel** [comment]
AWS has finally replied with something useful to try. However they are suggesting to update the License Manager config to stop data discovery. This is related to only one of the data syncs and it is to aid the investigation. However, if this license manager is in use, we would rather want to keep the sync in place. The next step is to identify if which license managers are actually in use (and there are still more data syncs than license managers in place).

**Comment by ewastempel** [comment]
I have followed the AWS advice, but I don't think the outcome was desired.
I have explained it all in the my response to the case, so I am attaching the content here:

> Hi Farley,
Thank you for your reply and your advice.
I have listened to your advice and I have disabled the Cross Account Resource Discovery on the License Manager using:
```
aws license-manager update-service-settings \
   --organization-configuration EnableIntegration=true \
   --no-enable-cross-accounts-discovery \
   --s3-bucket-arn arn:aws:s3:::aws-license-manager-service-643d94b3-abff-46cd-as
```
>and it worked fine (no errors) and the changes were reflected in the License Manager and the SSM resource data sync now has one less resource.
However, I have noticed that it in fact disabled the Cross Account Resource Discovery for the Self-managed licenses (please see the screenshot attached, this setting was enabled prior me running the cli command) and not the Managed licenses, which is not desired. You suggested that this change would affect the Managed License Manager setting and not the Self managed licenses, but it looks like it did the opposite.
We need the Cross Account Resource Discovery to be enabled, as we want to have visibility of all licenses (self managed and managed) managed centrally in this account, so I will be re-enabling this setting soon (the setting can be changed only once a day). 
Do you have more suggestions regarding this and do you have more insights on the other ssm resource data syncs please?
Thank you.
Ewa

**Comment by ewastempel** [comment]
I had a number of calls with AWS and this is the summary of findings so far:
```
1. license-manager-cross-account-discovery-for-del-admin-v2
- uses a zombie S3 bucket: aws-lm-bucket-245753150946-e889d8d4-6bcd-48cc-bae2-16f7bd5625c7
- originates from org security

157838100315 - account that owns the zombie data syncs
Role that run
arn:aws:iam::245753150946:role/aws-service-role/license-manager.member-account.amazonaws.com/AWSServiceRoleForAWSLicenseManagerMemberAccountRole

2. license-manager-cross-account-discovery-v2
- uses a zombie S3 bucket: aws-lm-bucket-295814833350-d952f1a5-3eb4-4746-a04c-3738069d6783
- originates from MoJ master account

157838100315 - account that owns the zombie data syncs
The role "AWSServiceRoleForAWSLicenseManagerMemberAccountRole" deleted and re-created the following data sync:
"license-manager-cross-account-discovery-v2
2025-03-14T14:10:49Z

3. license-manager-resource-data-sync-for-cross-account-discovery
- uses S3 bucket: aws-license-manager-service-54efe96a-05f8-4614-as (in MoJ Master)
- S3 referenced in the Licence Manager (in MoJ Master):  arn:aws:s3:::aws-license-manager-service-54efe96a-05f8-4614-as
- originates from MoJ master
- S3 has the AWS_License_Manager_Service_Metadata file with content: aws-lm-bucket-295814833350-d952f1a5-3eb4-4746-a04c-3738069d6783 (S3 bucket from data sync from 2.)

4. license-manager-cross-account-discovery-for-del-admin1710e486e9d
- uses S3 bucket: aws-license-manager-service-643d94b3-abff-46cd-as (in org security)
- S3 referenced in the License Manager (in org security):  arn:aws:s3:::aws-license-manager-service-643d94b3-abff-46cd-as
- originates from org security account
- S3 has the AWS_License_Manager_Service_Metadata file with content: aws-lm-bucket-245753150946-e889d8d4-6bcd-48cc-bae2-16f7bd5625c7 (S3 bucket from data sync from 1.)
```

AWS is further investigating the zombie resources and a possible bulk delete of them.

**Comment by ewastempel** [comment]
AWS has replied with their findings:
```
Investigation: 
==============
- I have updates regarding the origin of the syncs we looked at on call:
'''
license-manager-resource-data-sync-for-cross-account-discovery: This is created when the management account (MA) of the AWS Organization onboards to cross account discovery. It is expected for this data sync to remain there for cross account discovery functionality. 
license-manager-cross-account-discovery-v2: This is also created as part of MA onboarding. To keep the configuration updated this sync is deleted and recreated periodically. This is expected behavior. 
license-manager-cross-account-discovery-for-del-adminXXXX: This is created when the delegated admin (DA) of the AWS Organization onboards to cross account discovery. From looking at the case the delegated admin is 245753150946.
license-manager-cross-account-discovery-for-del-admin-v2: This is also created as part of DA onboarding.  To keep the configuration updated this sync is deleted and recreated periodically. This is expected behavior. 
'''
- As such it would follow that the above syncs are required for the cross account discovery. From checking internally all 4 syncs can be removed once both the Organizational Management Account and the Delegated Administrator account have Disabled Cross account discovery,
* There is an ongoing issue related to deleting the v2 syncs however this is being addressed internally.
- However, from discussing on call we would like to keep Cross Account discovery enabled. 
- I can also confrim that I have a 'license-manager-cross-account-discovery-v2' in my account aswell.
==============

Action Plan:
===============
- Considering the above information The 'cross-account-discovery' and the 'cross-account-discovery-v2' are expected. The reason for the other two del-admin syncs are because delegated administrator was configured which is expected behavior. 
- In the event you would like to keep using delegated administration I think the ideal course of action would be to request a resource data sync limit increase above 5. This would allow for additional syncs unrelated to license manager to be created while leaving your current license manager config untouched. 
- As mentioned on call you were hoping to clean up the resources as opposed to increase the limit. 
- If you would like to discuss my findings/next steps more in detail we can aim to connect over a live session once more. 
- Feel free to provide me a meeting time that works best for you and I will begin scheduling.
===============
```

It now appears that they can replicate the bahaviour on their side for all resource data syncs. This needs careful analysis on what is the best way forward to narrow down the number of the License Manager and data syncs.

A new [ticket](https://github.com/ministryofjustice/modernisation-platform/issues/9591) was open to track this work.

Closing this ticket.
